src/Services/Import/Savers/MySQLSaver.php

```diff
@@ -0,0 +1,143 @@
+        private TblproductdataRepository $productRepository,
+    )
+    {
+    }
+    
+    /** @var ImportRequest $request */
+    public function Save(array $requests): void
+    {
+        $this->products = [];
+        $this->requests = $requests;
+        $this->productsNames = $this->getProductsFieldsByObjMethod($this->requests, 'getProductName');
+
+        $this->setInvalidProductsWithExistsProductCode();
+        $this->setDiscontinued();
+
+        $this->setProducts();
+
+        $this->productRepository->saveProducts($this->products);
+    }
+
+    private function setInvalidProductsWithExistsProductCode() : void
+    {
+        $codes = $this->getProductsFieldsByObjMethod($this->requests, 'getProductCode');
+        $existsCodes = $this->productRepository->getExistsProductCodes($codes);
+
+        foreach($this->requests as $request) {
+            if($request->getIsValid()) {
+                if(array_search($request->getProductCode(), $existsCodes) !== false) {
+                    $request->setIsValid(false);
+                } else {
+                    $existsCodes[] = $request->getProductCode();
+                }
+            }
+        }
+    }
+
+    private function setDiscontinued() : void
+    {
+        $names = $this->getProductsFieldsByObjMethod($this->requests, 'getProductName');
+        $discontinuedProducts = $this->productRepository->getDiscontinuedProductsByNames($names);
+        $discontinuedProducts = $this->transformDiscontinuedArr($discontinuedProducts);
+
+        foreach($this->requests as $request) {
+            if($request->getIsValid()) {
+                if(array_key_exists($request->getProductName(), $discontinuedProducts)) {
+                    $request->setDiscontinuedDate($discontinuedProducts[$request->getProductName()]);
+                } else if ($request->getDiscontinued()) {
+                    $discontinuedProducts[$request->getProductName()] = $request->getDiscontinuedDate();
+                }
+            }
+        }
+    }
+
+    private function transformDiscontinuedArr(array $array) : array
+    {
+        $result = [];
+
+        for($i = 0; $i < count($array); $i++) {
+            $result[$array[$i]['strproductname']] = $array[$i]['dtmdiscontinued'];
+        }
+
+        return $result;
+    }
+    
+    private function setProducts() : void
+    {
+        $products = $this->getProducts();
+
+        /** @var Tblproductdata $product */
+        foreach($products as $product) {
+            $this->products[] = $product;
+        }
+    }
+
+    private function getProducts() : array
+    {
+        $products = [];
+
+        foreach($this->requests as $request) {
+            if($request->getIsValid()) {
+                $products[] = $this->createProduct($request);
+            }
+        }
+
+        return $products;
+    }
+
+    private function createProduct(ImportRequest $request) : Tblproductdata
+    {
+        $product = new Tblproductdata();
```

Comment:
if it's just a product why don't name entity Product?

Reply:
fixed